<a href="https://colab.research.google.com/github/Jubin5/Machine-Learning/blob/main/ML_Models(California_housing).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import fetch_california_housing
import pandas as pd

california_data = fetch_california_housing(as_frame=True)
df = california_data.frame
df


MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
0        -122.23        4.526  
1        -122.22        3.585  
2        -122.24        3.521  
3        -122.25        3.413  
4        -122.25        3.422  
...          ...          ...  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894  

[20640 rows x 9 columns]

In [ ]:
df.isnull().sum()
df.duplicated().sum()#From these two the output was 0 ,that means there is no mising values and no duplicate rows.

0

In [ ]:
df.select_dtypes(include=['number']).skew()


MedInc          1.646657
HouseAge        0.060331
AveRooms       20.697869
AveBedrms      31.316956
Population      4.935858
AveOccup       97.639561
Latitude        0.465953
Longitude      -0.297801
MedHouseVal     0.977763
dtype: float64

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_features = scaler.fit_transform(df)

scaled_df = pd.DataFrame(scaled_features, columns=df.columns)

print("\nSample of Scaled DataFrame:")
print(scaled_df.head())


Sample of Scaled DataFrame:
     MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  2.344766  0.982143  0.628559  -0.153758   -0.974429 -0.049597  1.052548   
1  2.332238 -0.607019  0.327041  -0.263336    0.861439 -0.092512  1.043185   
2  1.782699  1.856182  1.155620  -0.049016   -0.820777 -0.025843  1.038503   
3  0.932968  1.856182  0.156966  -0.049833   -0.766028 -0.050329  1.038503   
4 -0.012881  1.856182  0.344711  -0.032906   -0.759847 -0.085616  1.038503   

   Longitude  MedHouseVal  
0  -1.327835     2.129631  
1  -1.322844     1.314156  
2  -1.332827     1.258693  
3  -1.337818     1.165100  
4  -1.337818     1.172900  


In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
import numpy as np

X = scaled_df.drop(columns=['MedHouseVal'])
y = scaled_df['MedHouseVal']#Target variable

#Splitting the data

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

This algorithm works

In [ ]:
y_pred = model.predict(X_test)
y_pred

array([-1.16942746, -0.26391728,  0.55558107, ...,  2.08003582,
       -0.76351976, -0.05126459])

In [ ]:
#Step 6: Evaluate the model by linear regression
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Linear Regression Model Evaluation:")
print(f"Mean Squared Error: {mse}")
print(f"R-squared Score: {r2}")

Linear Regression Model Evaluation:
Mean Squared Error: 0.41747698052494986
R-squared Score: 0.575787706032451


This linear regression algorithm works by fitting a straight line in the data points with minimum error,Here the target variable is "MedHouseVal" and the other independent variables(x) are the other features excluding "MedHOuseVal"
The model has an error of 41% which shows that this model is better.

In [ ]:


from sklearn.tree import DecisionTreeRegressor

dt_model = DecisionTreeRegressor()
dt_model.fit(X_train, y_train)


DecisionTreeRegressor()

In [ ]:
dt_y_pred = dt_model.predict(X_test)


dt_mse = mean_squared_error(y_test, dt_y_pred)
dt_r2 = r2_score(y_test, dt_y_pred)

print("MSE:",dt_mse)
print("R2_score:",dt_r2)

MSE: 0.3700692560924254
R2_score: 0.6239602771476642


This is a Regression algorithm for supervised learning,In this model we will provide the x_train and y_train as the data set and when a data point coming from x_test this model will predict the coressponding y value with a small(37%)error only,And the r2score is 0.62.This model works by classifying the data in an inverted tree fashion

In [ ]:
from sklearn.ensemble import RandomForestRegressor
rf=RandomForestRegressor()
rf.fit(X_train,y_train)

RandomForestRegressor()

In [ ]:
rf_y_pred=rf.predict(X_test)
rf_y_pred

array([-1.36980634, -1.17055651,  2.44227442, ...,  2.41049696,
       -1.19008112, -0.39579427])

In [ ]:
rf_mse = mean_squared_error(y_test, rf_y_pred)
rf_r2 = r2_score(y_test, rf_y_pred)
print("MSE:",rf_mse)
print("R2_score:",rf_r2)

MSE: 0.1895254710658042
R2_score: 0.8074168430915434


This model works as same as that of the decision tree but the only difference is that it is a collection of decision trees such learning is known as ensemble learning ,so that the error is very minimum
Thatis: the error is very small (18%) and  the r2 score is 0.80

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor
gb_model = GradientBoostingRegressor()
gb_model.fit(X_train, y_train)

GradientBoostingRegressor()

In [ ]:
gb_y_pred = gb_model.predict(X_test)

gb_mse = mean_squared_error(y_test, gb_y_pred)
gb_r2 = r2_score(y_test, gb_y_pred)

print("\nGradient Boosting Regressor Model Evaluation:")
print("MSE:",gb_mse)
print("R2_score:",gb_r2)


Gradient Boosting Regressor Model Evaluation:
MSE: 0.22085568762773908
R2_score: 0.7755811643398038


This model builds a strong predictive model by combining multiple simple models (like decision trees) in a step-by-step process. Each new model focuses on correcting the errors made by the previous ones, and their predictions are added together to improve accuracy. This iterative approach continues until the errors are minimal or a set number of models are added.Here the error is minimum of 22% and an r2score of 0.77


In [ ]:

from sklearn.svm import SVR
svr_model = SVR()
svr_model.fit(X_train, y_train)

SVR()

In [ ]:
svr_y_pred = svr_model.predict(X_test)

svr_mse = mean_squared_error(y_test, svr_y_pred)
svr_r2 = r2_score(y_test, svr_y_pred)

print("\nSupport Vector Regressor Model Evaluation:")

print("MSe:",svr_mse)
print("R2_Score:",svr_r2)


Support Vector Regressor Model Evaluation:
MSe: 0.26439836363251595
R2_Score: 0.7313359979350702


This model works by finding the nearest data points and drawing a decision boundary,in regression, the "hyperplane" is not a decision boundary but a function (y=wx+b) that aims to approximate the target values within an ϵ margin
Here the error is 26% and th r2 score is 0.731.This algorithm is mainly used for classification so this is not apt for this data set.


And finally on looking through the metrices we we have the random forest regressor as the model with minimum error (18%) and its r2_score is 0.77.
The model with maximum error(41%) is the linear regression, so I prefer not to use that model